btf-rs/src/obj.rs

```diff
@@ -232,3 +239,177 @@ impl BtfBackend for CachedBtfObj {
+/// provides a faster initialization and lower memory footprint at the cost of
+/// slower API performances.
+struct MmapBtfObj {
+    endianness: cbtf::Endianness,
+    header: cbtf::btf_header,
+    // String offset from the base, 0 if not.
+    str_offset: u32,
+    // Type id offset from the base, 0 if not.
+    type_offset: u32,
+    // Number of types defined in the object.
+    types: usize,
+    // Memory-mapped reader.
+    mmap: Mmap,
+    // Map from type ids to their offsets in the mmaped BTF.
+    type_offsets: Vec<usize>,
+}
+
+impl MmapBtfObj {
+    fn from_mmap(mmap: Mmap, base: Option<Arc<BtfObj>>) -> Result<Self> {
+        let mut reader = Cursor::new(mmap);
+
+        // First parse the BTF header, retrieve the endianness & perform sanity
+        // checks.
+        let (header, endianness) = cbtf::btf_header::from_reader(&mut reader)?;
+        if header.version != 1 {
+            return Err(Error::Format(format!(
+                "Unsupported BTF version: {}",
+                header.version
+            )));
+        }
+
+        // Finally build our representation of the BTF types.
+        let offset = header.hdr_len + header.type_off;
+        reader.seek(SeekFrom::Start(offset as u64))?;
+
+        let mut offsets = Vec::new();
+        let mut types = 0;
+
+        let end_type_section = (offset + header.type_len) as u64;
+        while reader.stream_position()? < end_type_section {
+            offsets.push(reader.stream_position()? as usize);
+            cbtf::btf_skip_type(&mut reader, &endianness)?;
+            types += 1;
+        }
+
+        // Sanity check
+        if reader.stream_position()? != end_type_section {
+            return Err(Error::Format("Invalid type section".to_string()));
+        }
+
+        let (str_offset, type_offset) = match base {
+            Some(base) => (base.header_ref().str_len, base.types() as u32),
+            None => (0, 0),
+        };
+
+        Ok(Self {
+            endianness,
+            header,
+            str_offset,
+            type_offset,
+            types,
+            mmap: reader.into_inner(),
+            type_offsets: offsets,
+        })
+    }
+
+    // Iterate over the type names, calling a function on them (providing the
+    // type id and name bytes buffer).
+    fn iter_over_names<F>(&self, mut f: F) -> Result<()>
+    where
+        F: FnMut(u32, &[u8]) -> Result<()>,
+    {
+        let mmap = &self.mmap;
+
+        for (id, offset) in self.type_offsets.iter().enumerate() {
+            let bt = cbtf::btf_type::from_bytes(&mmap[*offset..], &self.endianness)?;
+            let name_off = match bt.name_offset() {
+                Some(offset) => offset,
+                None => continue,
+            };
+
+            if name_off < self.header.str_len {
+                let start = (self.header.hdr_len + self.header.str_off + name_off) as usize;
+
+                f(id as u32 + 1 + self.type_offset, &mmap[start..])?;
```

Comment:
Not hugely important but if the BTF file was corrupted and `start` would incorrectly be bigger than `mmap.len()` we would panic here. Panic-ing in library code is not good, so a quick boundary check and report the error would be better.

Reply:
Good catch! Maybe we should check the loaded BTF file is valid in `fn from_mmap` as it should not change under us later? This is already done for the BTF header and the type section, I can add a sanity check for the string section too.